src/core/indexer/mod.rs

```diff
@@ -0,0 +1,290 @@
+pub(crate) mod tables;
+
+use base64::Engine;
+
+use crate::{
+    core::indexer::tables::{BothSigCodex, CurrentSigCodex},
+    error::{Error, Result},
+};
+
+use super::util;
+
+/// Indexer is fully qualified cryptographic material primitive base class for
+/// indexed primitives. Indexed codes are a mix of indexed and variable length
+/// because code table has two char codes for compact variable length.
+pub(crate) struct Indexer {
+    /// stable (hard) part of derivation code
+    code: String,
+    /// unqualified crypto material usable for crypto operations
+    raw: Vec<u8>,
+    ///  main index offset into list or length of material
+    index: u32,
+    ///  other index offset into list or length of material
+    ondex: Option<u32>,
+}
+
+impl Default for Indexer {
+    fn default() -> Self {
+        Indexer {
+            raw: vec![],
+            code: tables::Codex::Ed25519_Sig.code().to_string(),
+            index: 0,
+            ondex: Some(u32::MAX),
+        }
+    }
+}
+
+impl Indexer {
+    pub fn new_with_code_and_raw(
+        code: &str,
+        raw: &[u8],
+        index: u32,
+        mut ondex: Option<u32>,
+    ) -> Result<Indexer> {
+        if code.is_empty() {
+            return Err(Box::new(Error::EmptyMaterial("empty code".to_string())));
+        }
+
+        let code = code.to_string();
+        let sizage_ = tables::sizage(&code)?;
+        if sizage_.fs == 0 {
+            return Err(Box::new(Error::InvalidVarRawSize(format!(
+                "unsupported variable size: code = [{}]",
```

Comment:
I actually removed all the square brackets in my most recent PR and surrounded strings with single quotes, that was really the only reason I started the square bracket shenanigans.

I'd now write this as:
```rust
"unsupported variable size: code = '{code}'"
```

(you also need to embed vars directly in the format strings where possible now, i think clippy changed)